paimon-core/src/main/java/org/apache/paimon/index/DynamicBucketIndexMaintainer.java

```diff
@@ -87,9 +89,11 @@ public List<IndexFileMeta> prepareCommit() {
                 throw new RuntimeException(e);
             }
             modified = false;
-            return Collections.singletonList(entry);
+            IndexFileMeta toRemove = beforeFile;
```

Comment:
In the streaming scenario, the index maintainer will always hold it and not create it repeatedly, so this logic needs to be added